### ISSUE TITLE
Review chapter 8 - Modular Development

### ISSUE THREAD

**Issue by wibblymat** [addyosmani/backbone-fundamentals#349]
Using chapter numbers in the issue titles doesn't seem so clever any more :)
1. Line 64:
   
   > When working with anonymous modules, the idea of a module's identity is DRY, making it trivial to avoid duplication of filenames and code.
   
   Not really sure what this sentence means
2. In the same section, the third parameter to `define()` is called the "definition function", but in the AMD spec says that it could just be an object. It is badly worded in the spec:
   
   >  factory, is a function that should be executed to instantiate the module or an object
   
   This is meant to be read as
   
   Factory is:
   - **a function that should be executed to instantiate the module**
     _or_
   - **an object**
   
   but on first glance you might think it meant:
   
   Factory is a function that should be executed to instantiate:
   - **the module** _or_
   - **an object**
   
   which I would guess is the source of the error. It is better explained in [the RequireJS docs](http://requirejs.org/docs/api.html#defsimple). Based on that I would avoid using the confusing language from the spec.
3. Line 131:
   
   > Addy's post on Writing Modular JS
   
   I know that this part was written by someone else but it does read like you are talking about yourself in the third person.
4. There needs to be a note somewhere that with RequireJS you don't need to put the ".js" on the end of script names.
5. Line 347 has sample code defining ItemCollectionView. Total nitpick, but shouldn't this be CartView? Would need updating in the referenced sample code on Github too so might not be worth the effort :)
6. Line 447:
   
   > I have yet to use it myself, but it comes as a recommended resource.
   
   I think we should be sounding a bit more authoritative.
7. Line 454:
   
   > Addy Osmani's screencast on build scripts
   
   Talking about yourself in the third-person again
8. Line 483:
   
   > Although the RequireJS optimizer works fine in both Node and Java environments, it's strongly recommended to run it under Node as it executes significantly faster there. In my experience, it's a piece of cake to get setup with either environment, so go for whichever you feel most comfortable with.
   
   Contradictory. Do you recommend going with whatever is most comfortable or do you recommend Node?
9. Line 487:
   
   > Below is an example of a build file taken from the modular jQuery Mobile app referenced later in this book.
   
   This example could be simpler and doesn't really need to mention that the app this file is for will be seen again later. I found the descriptions of the folder structure a little confusing, too. And why is r.js in `../../r.js` from the appDir? Shouldn't it be somewhere below the project root, not above it? In fact, if you `npm -g install requirejs` you can run it as just `r.js -o app.build.js` as it is placed in the PATH.
10. The section **Optimize and Build a Backbone.js JavaScript application with RequireJS using Packages** repeats a few introductory things from earlier in the chapter. No need to say what r.js is or what it is for as it has already been covered.
11. Same section, line 534:
    
    > When a JavaScript application is too complex or large to build in a single file, grouping the application’s components into packages allows for script dependencies to download in parallel, and facilitates only loading **packaged** and other modular code as the site experience requires the specific set of dependencies.
    
    Wasn't 100% sure what the second half of this sentence means. Reading further on it looks like this whole section needs a little bit of a language clean up.
12. In fact, this section is in need of a lot of rewording. There are a lot of sentences that don't really make sense. Since I don't have time left to do anything about it I'm going to skip it for this section, all the way down to **Exercise: Building a modular Backbone app with AMD & RequireJS**.
13. Line 663, in the build profile sample. Each module in the modules section seems to manually exclude each module that came before it. Surely this isn't the best way to do this? This looks pretty unmaintainable.
14. This whole example seems to contain a lot of things that aren't necessary to get across the point about building seperate packages. Pare down the amount of code in the samples to be just what is needed.
15. Line 1309: What is `obj`? In a requireJS module with an empty dependency list, wouldn't the first argument be `require`? Should the function that accepts `obj` be wrapped inside another function that actually defines the module?
16. Line 1916 (and similar on line 1946):
    
    > **Collection.doFakeFilter(filterFields, filterWords)** - returns the models count after fake-applying a call to `Collection.setFilter`.
    
    What does it mean to "fake-apply" a filter?
17. Line 1991 onwards appears to be garbage

**Comment by addyosmani** [comment]
cc @jackfranklin

Thanks a lot for the review, @wibblymat. The packages section does indeed also need a bit of a rewrite - did you feel it added enough value to warrant being in the book? Perhaps it would make sense to do as you said and pare it down to the bare minimum and maybe merge with the earlier section on AMD/RequireJS as an extension of the introduction.

**Comment by jackfranklin** [comment]
Thanks @wibblymat :) @addyosmani I'll happily take a look when I get some time over the parts of the review that apply to the bits I wrote and tweak them as required based on the feedback.

**Comment by addyosmani** [comment]
@jackfranklin Thank you, Jack. That would really help. I'll work out how we can firm up the other parts of this section that go into the topic of packages and so on. Go team!

**Comment by addyosmani** [comment]
Adding the review from @dcmaf to this thread to help aggregate the changes needed.

##### General

This chapter is very long (over 50 pages in the PDF version) and covers a wide range of subjects. It looks like some sections were originally separate chapters - I would recommend splitting them back out. For me, this is new material (I haven't used RequireJS) and I had a hard time getting through the chapter.

I would recommend grouping the AMD/RequireJS section and the modular Todo example into one chapter. It might even be possible to use the Todo example in place of the current discussion of how to use RequireJS with Backbone, though this may take more work than can be done at this point. I would suggest syncing the Todo example code with that in the earlier exercise so that the only differences are those introduced by the use of RequireJS.

The remaining sections of the chapter could either be standalone chapters or included in other chapters.

Part of what made the chapter difficult to get through was the redundancy due to the sections having been written separately. As a result, I found myself scanning paragraphs looking for what was new. I suspect this material would be much easier to digest if it were better integrated with the content of the chapters preceding it.

##### Detailed Comments

###### Section: Introduction

Second sentence of second paragraph: what does "It" refer to? Which specifications are being referenced?

>  Unlike some more traditional programming languages however, the current iteration of JavaScript (ECMA-262) doesn't provide developers with the means to import such modules of code in a clean, organized manner. It's one of the concerns with a specification that hasn't required great thought until more recent years when the need for more organized JavaScript applications has become apparent.

###### Section: Wrapping modules, views, and other components with AMD

In the example, is it necessary to pass 'underscore' in? It doesn't appear to be used in the model:

> ``` javascript
> define(['underscore', 'backbone'], function(_, Backbone) {
>   var myModel = Backbone.Model.extend({
> 
>    // Default attributes
>     defaults: {
>       content: 'hello world',
>     },
> 
>    // A dummy initialization method
>     initialize: function() {
>     },
> 
>    clear: function() {
>       this.destroy();
>       this.view.remove();
>     }
> 
>  });
>   return myModel;
> });
> ```

###### Section: Optimizing Backbone apps for production with the RequireJS Optimizer

Does the optimizer concatenate all the scripts into a single file? Does this lessen the benefit of using RequireJS/AMD to asynchronously load modules as needed? It looks like the next section addresses this, but it would be helpful to address here the benefits of optimizing to a single file since the introduction to AMD emphasized the benefits of loading files as-needed.

###### Section: Optimize and Build a Backbone.js JavaScript application with RequireJS using Packages
- Agree with Matt's comment on this section: needs work to integrate with the rest of the chapter/book and eliminate discussion which is redundant with content that has already been presented (make it more focused on the subject-at-hand).
- The three "Optimizing..." links to the RequireJS web site appear without any context (I thought they were references to upcoming subsections until I clicked on one of them):

> RequireJS, the (JavaScript) module loading library, has an [optimizer](http://requirejs.org/docs/optimization.html) to build a JavaScript-based application and provides various options. A build profile is the recipe for your build, much like a build.xml file is used to build a project with ANT. The benefit of building with **r.js** not only results in speedy script loading with minified code, but also provides a way to package components of your application.
> - [Optimizing one JavaScript file](http://requirejs.org/docs/optimization.html#onejs)
> - [Optimizing a whole project](http://requirejs.org/docs/optimization.html#wholeproject)
> - [Optimizing a project in layers or packages](http://requirejs.org/docs/faq-optimization.html#priority)
- I think there is a jump in complexity in this section which is overwhelming to a reader who is in the process of learning Backbone and RequireJS. Suggest moving it to later in the book or to an appendix, with a reference to that point from here.

###### Section: Build profile to optimize modular dependencies with code organized in packages

Second paragraph appears to be explaining the organization of the tree presented in the previous section, but is very dense. Needs to be broken into multiple paragraphs - possibly add annotations to the listing so the text can refer to specific parts of it.

> This strategy demonstrated builds common or site-wide groups of (core) _models_, _views_, collections which are extended from a base.js constructor which extends the appropriate backbone method, e.g. Backbone.Model. The _packages_ directory organizes code by section / responsibility, e.g. cart, checkout, etc. Notice that within the example _header_ package the directory structure is similar to the app root directory file structure. A _package_ (of modularized code) has dependencies from the common libraries in your application and also has specific code for the packages execution alone; other packages should not require another packages dependencies. A _utils_ directory has shims, helpers, and common library code to support the application. A _syncs_ directory to define persistence with your RESTful api and/or localStorage. The _vendor_ libraries folder will not be built, there is no need to do so, you may decide to use a CDN (then set these paths to : _[empty:](http://requirejs.org/docs/optimization.html#empty)_). And finally a _test_ directory for [Jasmine](http://pivotal.github.com/jasmine/) unit test specs, which may be ignored in the build as well if you choose.

###### Section: Exercise: Building a modular Backbone app with AMD & RequireJS
- First paragraph starts with "In this chapter". Is this intended to be a new chapter? It might make sense to make it one as this chapter is very long (even if the preceding section is pulled out):

> In this chapter, we'll look at our first practical Backbone & RequireJS project - how to build a modular Todo application. The application will allow us to add new todos, edit new todos and clear todo items that have been marked as completed. For a more advanced practical, see the section on mobile Backbone development.
- This section needs to be better integrated with the discussion of AMD/RequireJS earlier in the chapter as well as with the earlier Todo exercise. For example, in reading the discussion of the configuration object it is unclear to me whether this is new information or a restatement of the configuration concepts discussed earlier.

###### Section: Route-based Module Loading

Feels like this should be a separate chapter as well.

###### Section: Decoupling Backbone with the Mediator and Facade Patterns

Again, this feels like a completely new subject which belongs in its own chapter.

###### Section: Exercise
1. Under Mediator, references "aura/mediator.js". What is this a reference to?

> **Mediator**
> 
> Found in `aura/mediator.js`

###### Other
- What is best practice for organizing modules into files? With `<script>` tags, there is an incentive to minimize the number of files to keep the dependencies manageable. Is the number of files and their size not a concern when using RequireJS? For example, if you have a set of related views (such as ButtonView, ListBoxView, etc.) is it better to place them all in one file or should they each be in their own file? In Java-equivelent terms, should there be one file per package or one file per class?

**Comment by addyosmani** [comment]
@dcmaf re: 'Building a modular Backbone app with AMD & RequireJS' as it's own chapter - would it make sense to reposition this as another practical or should we keep it as just a separate chapter?

**Comment by addyosmani** [comment]
I'm also considering moving these two parts to the appendix. Wdyt?
- Module loader Router
- Using NodeJS to handle pushState

**Comment by addyosmani** [comment]
@dcmaf What I've done is broken the exercise, pagination and decoupling sections out into their own chapters. I feel this makes sense and provides readers with a clearer workflow for the book. 

I'm considering completely dropping "Optimize and Build a Backbone.js JavaScript application with RequireJS using Packages" - based on the feedback above, as I feel like it doesn't offer the beginner as much value. Does this make sense or would you rather see it moved to the appendix?

**Comment by addyosmani** [comment]
Just adding further questions:

I'm considering whether or not to drop the section on facades/mediators/permissions in favor of one that talks through just the mediator pattern and how to achieve this natively using Backbone 0.9.10. This itself can be done in just a few lines and I could probably re-use much of the mediator content itself from my JS patterns book.

Before I do this however I would like to ask, @wibblymat @dcmaf do you think this makes sense? Did you get anything of value out of the section on Aura? My reason for wanting to remove is that we've changed the APIs to it and don't feel it will stabilize before we get a chance to release this book.

**Comment by wibblymat** [comment]
@addyosmani I think you could safely drop the section. It does a good job of telling people why they should care about Aura and the basic concepts but has no direct relevance to Backbone. Also, the vast majority of Backbone applications will be simple enough that Aura is overkill.

**Comment by dcmaf** [comment]
@addyosmani Regarding the Module loader router, I suspect everything in that chapter from _Route-based module loading_ on belongs elsewhere (anything that is not part of the modular Todo exercise). Maybe an appendix on alternative loading and dependency management tools and strategies.

I'm okay with dropping the Packages section; I think it can be a useful example, but would require some rework to fit with the book since it has a very standalone feel. Might make a good online resource to reference.

I also agree with @wibblymat regarding the facades/mediators/permissions - it sounds like it may simplify the chapter by taking it out anyways.

Regarding partitioning the chapters into parts (which I think was brought up in issue #347), I could see having a part addressing real-world application issues, which would be a good place for the mediator pattern and the Marionette and Thorax extension chapters (not sure if it would make sense to include BBB/Grunt-BBB with them as well). A possible overall organization could be:
1. Introduction
- Part 1: Core Backbone
1. Fundamentals
2. Backbone Basics
3. Exercise 1: Todos
4. Exercise 2: Book Library
5. _(RESTful Sync chapter I can contribute could go here)_
6. Common Problems & Solutions
7. Modular Development
- Part 2: Unit Testing
1. Jasmine
2. QUnit
3. SinonJS
- Part 3: Backbone in the Real World
1. Backbone Boilerplate / Grunt-BBB
2. Paginating
3. Decoupling with Mediator
4. Marionette
5. Thorax
6. Mobile Applications
- Part 4: Final Thoughts
1. Resources
2. Conclusions
3. Appendix

With this organization, I can see a newbie getting enough out of parts 1/2 to get them started doing real development and then coming back to part 3 once they have enough experience to appreciate those sections. On the other hand, more experienced developers might want to skim through Part 1 and jump into Part 3. The order of parts 2 and 3 could be swapped; my thought is that the unit testing material would be relevant to newbies and is not dependent on any of the material in Part 3.

**Comment by jackfranklin** [comment]
Seeing as this is closed, just opened a new Pull Req to discuss further: https://github.com/addyosmani/backbone-fundamentals/pull/393
